### PR TITLE
Set Content-Type in headers of client http requests

### LINKED ISSUE #15758: Proposal: Add content-type on http-headers
Currently the command `kubectl -s http://192.168.0.15:8080 create -f sample.json` produces a packet similar to this one:
```
POST /api/v1/namespaces/default/pods HTTP/1.1
Host: 192.168.50.5:8080
User-Agent: kubectl/v1.0.6 (linux/amd64) kubernetes/388061f
Content-Length: 421
Accept-Encoding: gzip

{"kind":"Pod","apiVersion":"v1","metadata":{"name":"redis-controller","namespace":"default","creationTimestamp":null,"labels":{"app":"redis-controller"}},"spec":{"containers":[{"name":"redis","image":"redis","ports":[{"containerPort":6379,"protocol":"TCP"}],"resources":{},"terminationMessagePath":"/dev/termination-log","imagePullPolicy":"IfNotPresent"}],"restartPolicy":"Always","dnsPolicy":"ClusterFirst"},"status":{}}
```

Can someone add, if possible, the `Content-type: json` to the HTTP header of the `kubectl` request?

### PR DESCRIPTION
**Pull request by j3ffml**
Fixes #15758

cc @kubernetes/kubectl

### CODE REVIEW

**Comment by smarterclayton** [comment]
I don't think this is the right place to set this - it should be in the client/request.go path if we serialize to json only.  We don't always send JSON.

**Comment by jlowdermilk** [comment]
@smarterclayton, thanks, didn't know some paths are not json. I'll come back to this later.

**Comment by smarterclayton** [comment]
The code in request.go that sets r.body should also set a content-type, but
it can only be set if we're sure.  I think for now we can have anything
that goes through Codec to be assumed to be JSON on the way out, and for
the other endpoints just leave it unset.

File: pkg/client/unversioned/transport.go [hunk 1]
```diff
@@ -41,6 +41,7 @@ func (rt *userAgentRoundTripper) RoundTrip(req *http.Request) (*http.Response, e
 	}
 	req = cloneRequest(req)
 	req.Header.Set("User-Agent", rt.agent)
+	req.Header.Set("Content-Type", "application/json")
```
Comment (liggitt) [inline pkg/client/unversioned/transport.go]: I don't think this is always correct... @smarterclayton any opinion on this re: binary data?

Comment (liggitt) [inline pkg/client/unversioned/transport.go]: in any case, `userAgentRoundTripper` isn't the right place to set this

Comment (liggitt) [inline pkg/client/unversioned/transport.go]: won't it break for SPDY requests as well?

**Comment by jlowdermilk** [comment]
Moved the SetHeader call into `request.go` when body contains a serialized object and updated test. PTAL.

File: pkg/client/unversioned/transport.go (hunk 1 above)
Comment (jlowdermilk) [inline pkg/client/unversioned/transport.go]: Not sure if we have any requests that POST plain serialized json over SPDY, but I believe SPDY supports setting Content-Type header.

**Comment by jlowdermilk** [comment]
@k8s-bot test this

**Comment by jlowdermilk** [comment]
Ping @eparis, @smarterclayton. Any objections, or lgty?

**Comment by smarterclayton** [comment]
LGTM

File: pkg/client/unversioned/request.go [hunk 2]
```diff
@@ -491,6 +491,7 @@ func (r *Request) Body(obj interface{}) *Request {
 		}
 		glog.V(8).Infof("Request Body: %s", string(data))
 		r.body = bytes.NewBuffer(data)
+		r.SetHeader("Content-Type", "application/json")
```
Comment (lavalamp) [inline pkg/client/unversioned/request.go]: This is a property of the encoding, but I guess it's fine right now since we only have one.

**Pull request merged**